From 6015952363f071aacf6f5d2de607fc168daa1746 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Wed, 14 Aug 2024 19:34:26 +0800
Subject: [PATCH 25/46] grub-install: (loongarch64-efi) also install
 BOOTLOONGARCH.EFI

Some old-world firmware (especially that of the BPI01000 revision) does not
recognise the standardised BOOTLOONGARCH64.EFI and only sees BOOTLOONGARCH.EFI
as the removable image name. Make a copy (since it is FAT/VFAT) to satisfy
both old- and new-world firmware.
---
 util/grub-install.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/util/grub-install.c b/util/grub-install.c
index efa5cc573..6a5f86b09 100644
--- a/util/grub-install.c
+++ b/util/grub-install.c
@@ -938,6 +938,8 @@ main (int argc, char *argv[])
   const char *efi_distributor = NULL;
   const char *efi_file = NULL;
   const char *removeable_efi_file = NULL;
+  const char *efi_file_ow = NULL;
+  const char *removeable_efi_file_ow = NULL;
   char **grub_devices;
   grub_fs_t grub_fs;
   grub_device_t grub_dev = NULL;
@@ -1274,6 +1276,7 @@ main (int argc, char *argv[])
 	      break;
 	    case GRUB_INSTALL_PLATFORM_LOONGARCH64_EFI:
 	      efi_file = "BOOTLOONGARCH64.EFI";
+	      efi_file_ow = "BOOTLOONGARCH.EFI";
 	      break;
 	    case GRUB_INSTALL_PLATFORM_RISCV32_EFI:
 	      efi_file = "BOOTRISCV32.EFI";
@@ -1288,6 +1291,7 @@ main (int argc, char *argv[])
 	  if (force_extra_removable)
 	    {
 	      removeable_efi_file = efi_file;
+	      removeable_efi_file_ow = efi_file_ow;
 	    }
 	}
       if (!removable)
@@ -1314,6 +1318,7 @@ main (int argc, char *argv[])
 	      break;
 	    case GRUB_INSTALL_PLATFORM_LOONGARCH64_EFI:
 	      efi_file = "grubloongarch64.efi";
+	      efi_file_ow = NULL;
 	      break;
 	    case GRUB_INSTALL_PLATFORM_RISCV32_EFI:
 	      efi_file = "grubriscv32.efi";
@@ -2091,8 +2096,21 @@ main (int argc, char *argv[])
 	char *dst = grub_util_path_concat (2, efidir, efi_file);
 	grub_install_copy_file (imgfile, dst, 1);
 
+	if (platform == GRUB_INSTALL_PLATFORM_LOONGARCH64_EFI && efi_file_ow)
+	  {
+	    char *dst_ow = grub_util_path_concat (2, efidir, efi_file_ow);
+	    grub_install_copy_file (imgfile, dst_ow, 1);
+	    free (dst_ow);
+	  }
+
 	if (force_extra_removable)
+	  {
 	    also_install_removable(imgfile, base_efidir, removeable_efi_file);
+	    if (platform == GRUB_INSTALL_PLATFORM_LOONGARCH64_EFI && removeable_efi_file_ow)
+	      {
+		also_install_removable(imgfile, base_efidir, removeable_efi_file_ow);
+	      }
+	  }
 
 	grub_set_install_backup_ponr ();
 
-- 
2.47.0

